Check minimal quantity

sadlyblue opened the pull request:
Solve #10162. After the product is added to the cart, it will return an error if the customer tries to decrease the product quantity below the minimal_quantity (for the product or combination).

<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | issue #10162 from github
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10162.
| How to test?  | As indicated in the issue, Have a product (or combination) with minimal_quantity > 1. Add to cart the minimal_quantity. With this fix, when you try to remove below the minimal_quantity it will display/return an error message.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10279)
<!-- Reviewable:end -->

sadlyblue committed:
Check minimal quantity

Solve #10162. After the product is added to the cart, it will return an error if the customer tries to decrease the product quantity below the minimal_quantity (for the product or combination).

LouiseBonnard commented on controllers/front/CartController.php:
```diff
@@ -432,6 +432,35 @@ protected function processChangeProductInCart()
+                 array_push(
+                     $this->{$ErrorKey},
+                     $this->trans(
+                         'The minimum purchase order quantity for the product %product% is %quantity%.',
```
Is it necessary to add the product name here? I mean, it is already mentioned on the product page... perhaps we can just keep the already translation key _The minimum purchase order quantity for the product is %quantity%._ in _Shop.Theme.Checkout_?

jolelievre replied:
hello, could you please keep the `$this->{$ErrorKey}[]` notation? not that I dislike `array_push` but for consistency with rest of the code in this method

PierreRambaud replied:
More array_push have poor performance when used with one element.

sadlyblue replied:
Saw both being used in the previous version (the one I tested on, even before the file moved to Presenter folder if i'm not mistaken). But I agree, no need for array_push.

LouiseBonnard commented on controllers/front/CartController.php:
```diff
@@ -432,6 +432,35 @@ protected function processChangeProductInCart()
+                 array_push(
+                     $this->{$ErrorKey},
+                     $this->trans(
+                         'The minimum purchase order quantity for the product %product% is %quantity%.',
```
Same here!

sadlyblue replied:
Not required. But this also triggers if the customer adjust the quantity in the cart/checkout process. This way would return the product name to better identify the product. In case of slower systems the user could try to adjust more than one quantiy without waiting for the result. And since some other warnings have the product name. But agree, not required.

LouiseBonnard replied:
Well, you're right, it could be useful to have it notified during the checkout. Can you add it to this PR? Or it could just be displayed next to the item concerned so that we don't have to change the wording, what do you think?

jolelievre replied:
same comment as above `$this->{$ErrorKey}[]`

PierreRambaud commented on controllers/front/CartController.php:
```diff
@@ -432,6 +432,35 @@ protected function processChangeProductInCart()
+                 return;
+            }
+        }
+        else{
```
Can you execute cs fixer?

sadlyblue committed:
Array_push changed

array_push to $this->{$ErrorKey}[]

matks committed:
Apply php-cs-fixer to comply with CS rules

matks commented:
@sadlyblue I applied php-cs-fixer for you ;) now waiting for travis build, it should be green ...

marionf commented:
Hello @sadlyblue 

If I set the minimum quantity to 4 and try to decrease in cart the quantity, I have the error message and can't order but the quantity is still on 4, it should be on 3 with the order button disabled
Then if I increase I have 5 but I should have 4

https://drive.google.com/open?id=1ZueXDVq8BBup8DnA6B92tcjVNBDGlIvT

Also, if the product has combinations, it should be good to have the combination displayed in the error message

sadlyblue commented:
Hi @marionf ,
You are right, since it returns an error, it doesn't unlock the checkout button.
Will have a look at it.
Is it supposed to not let the order finish and show the amount less than the minimum, or not let the amount be less than the minimum amount?
I would vote for not letting the amount be less than the minimum amout, only needing to "activate" the checkout button. To have the same behaviour as in the product page (doesn't let adding to cart less than the minimum amount.

sadlyblue committed:
Fix checkout button unavailable

Changing $this->{$ErrorKey} to $this->errors returns the same error, but the checkout button is enabled (as per @marionf comment and bug reported).
I believe it should not decrease the product quantity, since in the product page we con only add the minimum amount or can't add at all.

marionf commented:
Seems better like this :)
Thank you @sadlyblue

PierreRambaud commented:
Thanks @sadlyblue
